1840: check for concurrent edits from database grade events

payten opened the pull request:
Hi @steveswinsburg,

I've had a go at refactoring the `isAnotherUserEditing` API to be cluster friendlier by moving the check from notifications in memory to instead hit the database and return any `GradingEvent` made after a particular timestamp.

This is how it works:
1. Add new service method to return all grading events for a set of assignments after a given timestamp
2. Mark the time the grade matrix was build and make it known to the javascript
3. The JavaScript polls `isAnotherUserEditing` with the timestamp
4. `isAnotherUserEditing` hits the new service API `getGradingEvents` to view any events made after the given timestamp.
5. We filter the events to those made by other users i.e. not the current user
6. And lastly, morph those `GradingEvent` objects into the `GbGradeCell` model as expected by the `isAnotherUserEditing` service and JavaScript.

I've made the `getGradingEvents` take a list of the assignments, which I think makes it TA friendly - so only changes for assignments the TA can see are returned via the `isAnotherUserEditing` entity service call.

Does this all seem ok?

Thanks,
Payten

Delivers #1840.

steveswinsburg commented on edu-services/gradebook-service/impl/src/java/org/sakaiproject/component/gradebook/GradebookServiceHibernateImpl.java:
```diff
@@ -3511,4 +3511,35 @@ public Object doInHibernate(Session session) throws HibernateException {
+		HibernateCallback hc = new HibernateCallback() {
+			@Override
+			public Object doInHibernate(Session session) throws HibernateException, SQLException {
+				Query q = session.createQuery("from GradingEvent as ge where ge.dateGraded >= :since" +
```
Will likely need an index on these two columns for an efficient lookup, which may be a combination of both fields. But need to profile with lots of rows. Are you able to load up the db then do an explain plan on this?
````
SELECT
* 
FROM gb_grading_event_t ge
WHERE ge.date_graded >= :since
AND ge.gradable_object_id in (:assignment_ids);
````

Alternative, grab the actual generated SQL by turinng on the hibernate show sql feature in sakai.properties, just wondering if that gradeableObject.id does an implicit join.

payten replied:
Sure thing.  I'll check it out.

steveswinsburg commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/business/GradebookNgBusinessService.java:
```diff
@@ -1204,83 +1183,39 @@ public int compare(final User u1, final User u2) {
+		final List<GradingEvent> events = this.gradebookService.getGradingEvents(assignmentIds, since);
+
+		// filter out any events made by the current user
+		final Map<String, GbUser> userCache = new HashMap<>();
```
is using a standard HashMap for this sufficient, does this need to be cluster safe?

payten replied:
Yeah, this was just to ensure we only need to hit `getUser` once for each distinct user as we iterate over the events - this "cache" is per-request/method call so doesn't need to be cluster safe... perhaps the `userCache` attribute name is a little misleading!

steveswinsburg replied:
OK got it, yeah the cache bit tricked me. Maybe userCheck would be better hehe.

steveswinsburg commented:
My main concern about this is the number of db reads increases significantly.

Could you turn on `hibernate.show_sql=true` and capture the DB reads whenever this is invoked? Maybe we can tune some stuff in GBNG business to reduce this.

payten commented:
Sure thing.  I'll observe the exact DB reads and report back.

payten commented:
I've just pushed up a patch.. and with that each call to `isAnotherUserEditing` invokes 12 SQL queries; 6 to return the assignments for current user and another 6 to return the grading events since a particular timestamp.  Here's a list of the SQL commands when there is one GradeEvent returned:

`getViewableAssignmentsForCurrentUser`:
```
Hibernate: select gradebook0_.ID as ID15_, gradebook0_.VERSION as VERSION15_, gradebook0_.GRADEBOOK_UID as GRADEBOOK3_15_, gradebook0_.NAME as NAME15_, gradebook0_.SELECTED_GRADE_MAPPING_ID as SELECTED5_15_, gradebook0_.ASSIGNMENTS_DISPLAYED as ASSIGNME6_15_, gradebook0_.COURSE_GRADE_DISPLAYED as COURSE7_15_, gradebook0_.COURSE_POINTS_DISPLAYED as COURSE8_15_, gradebook0_.TOTAL_POINTS_DISPLAYED as TOTAL9_15_, gradebook0_.COURSE_AVERAGE_DISPLAYED as COURSE10_15_, gradebook0_.ALL_ASSIGNMENTS_ENTERED as ALL11_15_, gradebook0_.LOCKED as LOCKED15_, gradebook0_.GRADE_TYPE as GRADE13_15_, gradebook0_.CATEGORY_TYPE as CATEGORY14_15_, gradebook0_.IS_EQUAL_WEIGHT_CATS as IS15_15_, gradebook0_.IS_SCALED_EXTRA_CREDIT as IS16_15_, gradebook0_.DO_SHOW_MEAN as DO17_15_, gradebook0_.DO_SHOW_MEDIAN as DO18_15_, gradebook0_.DO_SHOW_MODE as DO19_15_, gradebook0_.DO_SHOW_RANK as DO20_15_, gradebook0_.DO_SHOW_ITEM_STATS as DO21_15_, gradebook0_.DO_SHOW_STATISTICS_CHART as DO22_15_, gradebook0_.COURSE_LETTER_GRADE_DISPLAYED as COURSE23_15_ from GB_GRADEBOOK_T gradebook0_ where gradebook0_.GRADEBOOK_UID=?
Hibernate: select grademappi0_.ID as ID23_3_, grademappi0_.VERSION as VERSION23_3_, grademappi0_.GRADEBOOK_ID as GRADEBOOK4_23_3_, grademappi0_.GB_GRADING_SCALE_T as GB5_23_3_, grademappi0_.OBJECT_TYPE_ID as OBJECT2_23_3_, gradebook1_.ID as ID15_0_, gradebook1_.VERSION as VERSION15_0_, gradebook1_.GRADEBOOK_UID as GRADEBOOK3_15_0_, gradebook1_.NAME as NAME15_0_, gradebook1_.SELECTED_GRADE_MAPPING_ID as SELECTED5_15_0_, gradebook1_.ASSIGNMENTS_DISPLAYED as ASSIGNME6_15_0_, gradebook1_.COURSE_GRADE_DISPLAYED as COURSE7_15_0_, gradebook1_.COURSE_POINTS_DISPLAYED as COURSE8_15_0_, gradebook1_.TOTAL_POINTS_DISPLAYED as TOTAL9_15_0_, gradebook1_.COURSE_AVERAGE_DISPLAYED as COURSE10_15_0_, gradebook1_.ALL_ASSIGNMENTS_ENTERED as ALL11_15_0_, gradebook1_.LOCKED as LOCKED15_0_, gradebook1_.GRADE_TYPE as GRADE13_15_0_, gradebook1_.CATEGORY_TYPE as CATEGORY14_15_0_, gradebook1_.IS_EQUAL_WEIGHT_CATS as IS15_15_0_, gradebook1_.IS_SCALED_EXTRA_CREDIT as IS16_15_0_, gradebook1_.DO_SHOW_MEAN as DO17_15_0_, gradebook1_.DO_SHOW_MEDIAN as DO18_15_0_, gradebook1_.DO_SHOW_MODE as DO19_15_0_, gradebook1_.DO_SHOW_RANK as DO20_15_0_, gradebook1_.DO_SHOW_ITEM_STATS as DO21_15_0_, gradebook1_.DO_SHOW_STATISTICS_CHART as DO22_15_0_, gradebook1_.COURSE_LETTER_GRADE_DISPLAYED as COURSE23_15_0_, grademappi2_.ID as ID23_1_, grademappi2_.VERSION as VERSION23_1_, grademappi2_.GRADEBOOK_ID as GRADEBOOK4_23_1_, grademappi2_.GB_GRADING_SCALE_T as GB5_23_1_, grademappi2_.OBJECT_TYPE_ID as OBJECT2_23_1_, gradingsca3_.ID as ID19_2_, gradingsca3_.VERSION as VERSION19_2_, gradingsca3_.SCALE_UID as SCALE4_19_2_, gradingsca3_.NAME as NAME19_2_, gradingsca3_.UNAVAILABLE as UNAVAILA6_19_2_ from GB_GRADE_MAP_T grademappi0_ inner join GB_GRADEBOOK_T gradebook1_ on grademappi0_.GRADEBOOK_ID=gradebook1_.ID left outer join GB_GRADE_MAP_T grademappi2_ on gradebook1_.SELECTED_GRADE_MAPPING_ID=grademappi2_.ID left outer join GB_GRADING_SCALE_T gradingsca3_ on grademappi0_.GB_GRADING_SCALE_T=gradingsca3_.ID where grademappi0_.ID=?
Hibernate: select defaultbot0_.GRADING_SCALE_ID as GRADING1_19_0_, defaultbot0_.PERCENT as PERCENT0_, defaultbot0_.LETTER_GRADE as LETTER3_0_ from GB_GRADING_SCALE_PERCENTS_T defaultbot0_ where defaultbot0_.GRADING_SCALE_ID=?
Hibernate: select grades0_.GRADING_SCALE_ID as GRADING1_19_0_, grades0_.LETTER_GRADE as LETTER2_0_, grades0_.GRADE_IDX as GRADE3_0_ from GB_GRADING_SCALE_GRADES_T grades0_ where grades0_.GRADING_SCALE_ID=?
Hibernate: select grademap0_.GRADE_MAP_ID as GRADE1_23_0_, grademap0_.PERCENT as PERCENT0_, grademap0_.LETTER_GRADE as LETTER3_0_ from GB_GRADE_TO_PERCENT_MAPPING_T grademap0_ where grademap0_.GRADE_MAP_ID=?
Hibernate: select assignment0_.ID as ID16_, assignment0_.VERSION as VERSION16_, assignment0_.GRADEBOOK_ID as GRADEBOOK4_16_, assignment0_.NAME as NAME16_, assignment0_.REMOVED as REMOVED16_, assignment0_.SORT_ORDER as SORT7_16_, assignment0_.CATEGORIZED_SORT_ORDER as CATEGORI8_16_, assignment0_.POINTS_POSSIBLE as POINTS9_16_, assignment0_.DUE_DATE as DUE10_16_, assignment0_.NOT_COUNTED as NOT11_16_, assignment0_.EXTERNALLY_MAINTAINED as EXTERNALLY12_16_, assignment0_.EXTERNAL_STUDENT_LINK as EXTERNAL13_16_, assignment0_.EXTERNAL_INSTRUCTOR_LINK as EXTERNAL14_16_, assignment0_.EXTERNAL_ID as EXTERNAL15_16_, assignment0_.EXTERNAL_APP_NAME as EXTERNAL16_16_, assignment0_.IS_EXTRA_CREDIT as IS17_16_, assignment0_.ASSIGNMENT_WEIGHTING as ASSIGNMENT18_16_, assignment0_.RELEASED as RELEASED16_, assignment0_.CATEGORY_ID as CATEGORY20_16_, assignment0_.UNGRADED as UNGRADED16_, assignment0_.IS_NULL_ZERO as IS22_16_, assignment0_.HIDE_IN_ALL_GRADES_TABLE as HIDE23_16_ from GB_GRADABLE_OBJECT_T assignment0_ where assignment0_.OBJECT_TYPE_ID=1 and assignment0_.GRADEBOOK_ID=? and assignment0_.REMOVED=0
```

And then `getGradingEvents`:
```
Hibernate: select gradingeve0_.ID as ID18_, gradingeve0_.GRADABLE_OBJECT_ID as GRADABLE2_18_, gradingeve0_.GRADER_ID as GRADER3_18_, gradingeve0_.STUDENT_ID as STUDENT4_18_, gradingeve0_.DATE_GRADED as DATE5_18_, gradingeve0_.GRADE as GRADE18_ from GB_GRADING_EVENT_T gradingeve0_ where gradingeve0_.DATE_GRADED>=? and (gradingeve0_.GRADABLE_OBJECT_ID in (? , ? , ? , ?))
Hibernate: select gradableob0_.ID as ID16_4_, gradableob0_.VERSION as VERSION16_4_, gradableob0_.GRADEBOOK_ID as GRADEBOOK4_16_4_, gradableob0_.NAME as NAME16_4_, gradableob0_.REMOVED as REMOVED16_4_, gradableob0_.SORT_ORDER as SORT7_16_4_, gradableob0_.CATEGORIZED_SORT_ORDER as CATEGORI8_16_4_, gradableob0_.POINTS_POSSIBLE as POINTS9_16_4_, gradableob0_.DUE_DATE as DUE10_16_4_, gradableob0_.NOT_COUNTED as NOT11_16_4_, gradableob0_.EXTERNALLY_MAINTAINED as EXTERNALLY12_16_4_, gradableob0_.EXTERNAL_STUDENT_LINK as EXTERNAL13_16_4_, gradableob0_.EXTERNAL_INSTRUCTOR_LINK as EXTERNAL14_16_4_, gradableob0_.EXTERNAL_ID as EXTERNAL15_16_4_, gradableob0_.EXTERNAL_APP_NAME as EXTERNAL16_16_4_, gradableob0_.IS_EXTRA_CREDIT as IS17_16_4_, gradableob0_.ASSIGNMENT_WEIGHTING as ASSIGNMENT18_16_4_, gradableob0_.RELEASED as RELEASED16_4_, gradableob0_.CATEGORY_ID as CATEGORY20_16_4_, gradableob0_.UNGRADED as UNGRADED16_4_, gradableob0_.IS_NULL_ZERO as IS22_16_4_, gradableob0_.HIDE_IN_ALL_GRADES_TABLE as HIDE23_16_4_, gradableob0_.OBJECT_TYPE_ID as OBJECT2_16_4_, gradebook1_.ID as ID15_0_, gradebook1_.VERSION as VERSION15_0_, gradebook1_.GRADEBOOK_UID as GRADEBOOK3_15_0_, gradebook1_.NAME as NAME15_0_, gradebook1_.SELECTED_GRADE_MAPPING_ID as SELECTED5_15_0_, gradebook1_.ASSIGNMENTS_DISPLAYED as ASSIGNME6_15_0_, gradebook1_.COURSE_GRADE_DISPLAYED as COURSE7_15_0_, gradebook1_.COURSE_POINTS_DISPLAYED as COURSE8_15_0_, gradebook1_.TOTAL_POINTS_DISPLAYED as TOTAL9_15_0_, gradebook1_.COURSE_AVERAGE_DISPLAYED as COURSE10_15_0_, gradebook1_.ALL_ASSIGNMENTS_ENTERED as ALL11_15_0_, gradebook1_.LOCKED as LOCKED15_0_, gradebook1_.GRADE_TYPE as GRADE13_15_0_, gradebook1_.CATEGORY_TYPE as CATEGORY14_15_0_, gradebook1_.IS_EQUAL_WEIGHT_CATS as IS15_15_0_, gradebook1_.IS_SCALED_EXTRA_CREDIT as IS16_15_0_, gradebook1_.DO_SHOW_MEAN as DO17_15_0_, gradebook1_.DO_SHOW_MEDIAN as DO18_15_0_, gradebook1_.DO_SHOW_MODE as DO19_15_0_, gradebook1_.DO_SHOW_RANK as DO20_15_0_, gradebook1_.DO_SHOW_ITEM_STATS as DO21_15_0_, gradebook1_.DO_SHOW_STATISTICS_CHART as DO22_15_0_, gradebook1_.COURSE_LETTER_GRADE_DISPLAYED as COURSE23_15_0_, grademappi2_.ID as ID23_1_, grademappi2_.VERSION as VERSION23_1_, grademappi2_.GRADEBOOK_ID as GRADEBOOK4_23_1_, grademappi2_.GB_GRADING_SCALE_T as GB5_23_1_, grademappi2_.OBJECT_TYPE_ID as OBJECT2_23_1_, category3_.ID as ID27_2_, category3_.VERSION as VERSION27_2_, category3_.GRADEBOOK_ID as GRADEBOOK3_27_2_, category3_.NAME as NAME27_2_, category3_.WEIGHT as WEIGHT27_2_, category3_.DROP_LOWEST as DROP6_27_2_, category3_.REMOVED as REMOVED27_2_, category3_.IS_EXTRA_CREDIT as IS8_27_2_, category3_.IS_EQUAL_WEIGHT_ASSNS as IS9_27_2_, category3_.IS_UNWEIGHTED as IS10_27_2_, category3_.CATEGORY_ORDER as CATEGORY11_27_2_, category3_.ENFORCE_POINT_WEIGHTING as ENFORCE12_27_2_, category3_.DROP_HIGHEST as DROP13_27_2_, category3_.KEEP_HIGHEST as KEEP14_27_2_, gradebook4_.ID as ID15_3_, gradebook4_.VERSION as VERSION15_3_, gradebook4_.GRADEBOOK_UID as GRADEBOOK3_15_3_, gradebook4_.NAME as NAME15_3_, gradebook4_.SELECTED_GRADE_MAPPING_ID as SELECTED5_15_3_, gradebook4_.ASSIGNMENTS_DISPLAYED as ASSIGNME6_15_3_, gradebook4_.COURSE_GRADE_DISPLAYED as COURSE7_15_3_, gradebook4_.COURSE_POINTS_DISPLAYED as COURSE8_15_3_, gradebook4_.TOTAL_POINTS_DISPLAYED as TOTAL9_15_3_, gradebook4_.COURSE_AVERAGE_DISPLAYED as COURSE10_15_3_, gradebook4_.ALL_ASSIGNMENTS_ENTERED as ALL11_15_3_, gradebook4_.LOCKED as LOCKED15_3_, gradebook4_.GRADE_TYPE as GRADE13_15_3_, gradebook4_.CATEGORY_TYPE as CATEGORY14_15_3_, gradebook4_.IS_EQUAL_WEIGHT_CATS as IS15_15_3_, gradebook4_.IS_SCALED_EXTRA_CREDIT as IS16_15_3_, gradebook4_.DO_SHOW_MEAN as DO17_15_3_, gradebook4_.DO_SHOW_MEDIAN as DO18_15_3_, gradebook4_.DO_SHOW_MODE as DO19_15_3_, gradebook4_.DO_SHOW_RANK as DO20_15_3_, gradebook4_.DO_SHOW_ITEM_STATS as DO21_15_3_, gradebook4_.DO_SHOW_STATISTICS_CHART as DO22_15_3_, gradebook4_.COURSE_LETTER_GRADE_DISPLAYED as COURSE23_15_3_ from GB_GRADABLE_OBJECT_T gradableob0_ inner join GB_GRADEBOOK_T gradebook1_ on gradableob0_.GRADEBOOK_ID=gradebook1_.ID left outer join GB_GRADE_MAP_T grademappi2_ on gradebook1_.SELECTED_GRADE_MAPPING_ID=grademappi2_.ID left outer join GB_CATEGORY_T category3_ on gradableob0_.CATEGORY_ID=category3_.ID left outer join GB_GRADEBOOK_T gradebook4_ on category3_.GRADEBOOK_ID=gradebook4_.ID where gradableob0_.ID=?
Hibernate: select gradingsca0_.ID as ID19_0_, gradingsca0_.VERSION as VERSION19_0_, gradingsca0_.SCALE_UID as SCALE4_19_0_, gradingsca0_.NAME as NAME19_0_, gradingsca0_.UNAVAILABLE as UNAVAILA6_19_0_ from GB_GRADING_SCALE_T gradingsca0_ where gradingsca0_.ID=?
Hibernate: select grademap0_.GRADE_MAP_ID as GRADE1_23_0_, grademap0_.PERCENT as PERCENT0_, grademap0_.LETTER_GRADE as LETTER3_0_ from GB_GRADE_TO_PERCENT_MAPPING_T grademap0_ where grademap0_.GRADE_MAP_ID=?
Hibernate: select defaultbot0_.GRADING_SCALE_ID as GRADING1_19_0_, defaultbot0_.PERCENT as PERCENT0_, defaultbot0_.LETTER_GRADE as LETTER3_0_ from GB_GRADING_SCALE_PERCENTS_T defaultbot0_ where defaultbot0_.GRADING_SCALE_ID=?
Hibernate: select grades0_.GRADING_SCALE_ID as GRADING1_19_0_, grades0_.LETTER_GRADE as LETTER2_0_, grades0_.GRADE_IDX as GRADE3_0_ from GB_GRADING_SCALE_GRADES_T grades0_ where grades0_.GRADING_SCALE_ID=?
```

It looks like, correctly, there is only one query against the `GB_GRADING_EVENT_T` table without any inline joins - hurray!  However, hibernate proceeds to expand the gradable_object, which isn't required for this particular workflow.  Is there are way we can avoid that?  

I'm sure we could create some our own SQL query to avoid hitting the `getAssignments` etc, but that would need to bypass Hibernate and it's *expensive* magic.

And you're right, we'd need to add some indexes to the GB_GRADING_EVENT_T table to ensure it doesn't chug, currently there's only indexes on the id and foreign key (gradable_object_id).

steveswinsburg commented:
OK those first 5 are from the call to getGradebook as a result of the getAssignments call. I am wondering if we can cache the gradebook object, or whether its worthwhile. If we did then we could save a heap of DB hits, just wondering if its worth the complexity.

How about we raise a separate ticket to improve the performance of getGradebookEvents without expanding the rest of the data (I think if we use plain HQL it would work) and we can look at it later.

I checked the explain plan and it's using the index 💛

steveswinsburg commented:
If this is good to go I can merge it, was there anything else to add?

payten commented:
I've just updated the `userCache` attribute and this should be right go.  Thanks @steveswinsburg 🍻

steveswinsburg commented:
Ok, here we go!
